packages/@aws-cdk/aws-apigateway/lib/authorizer.ts

```diff
@@ -1,9 +1,71 @@
+import { Construct, Lazy, Resource, ResourceProps } from '@aws-cdk/core';
+import { IRestApi } from '../lib/restapi';
+
+const CUSTOM_AUTH_SYMBOL = Symbol.for('@aws-cdk/apigateway.CustomAuthorizer');
+
 /**
  * Represents an API Gateway authorizer.
  */
 export interface IAuthorizer {
   /**
    * The authorizer ID.
+   * @attribute
    */
   readonly authorizerId: string;
+}
+
+/**
+ * Base class for all custom authorizers
+ */
+export abstract class CustomAuthorizer extends Resource implements IAuthorizer {
+
+  /**
+   * Check if the given object is of type CustomAuthorizer
+   */
+  public static isCustomAuthorizer(x: any): x is CustomAuthorizer {
+    return x !== null && typeof(x) === 'object' && CUSTOM_AUTH_SYMBOL in x;
+  }
+
+  /**
+   * The authorizer ID.
+   * @attribute
+   */
+  public abstract readonly authorizerId: string;
+
+  private _restApi?: IRestApi;
+
+  constructor(scope: Construct, id: string, props?: ResourceProps) {
+    super(scope, id, props);
+
+    Object.defineProperty(this, CUSTOM_AUTH_SYMBOL, { value: true });
+  }
+
+  /**
+   * Associate this authorizer to a specific {@link RestApi}.
+   * Typically, this is taken care of by the module while defining {@link Method}s via the {@link ResourceBase#addMethod | addMethod()} API call.
+   *
+   * @throws when a different RestApi is already associated to this authorizer.
+   */
+  public set restApi(restApi: IRestApi) {
+    if (this._restApi && this._restApi !== restApi) {
+      throw new Error(`Cannot attach authorizer [${this.node.uniqueId}] to RestApi [${restApi.node.uniqueId}]. `
+        + `Already attached to RestApi [${this._restApi.node.uniqueId}]`);
+    }
+    this._restApi = restApi;
+  }
+
+  /**
+   * Returns a lazy that will resolve to the restApiId at the time of synthesis.
+   */
+  protected get restApiId(): string {
```

Comment:
Assign in the ctor, no need for a getter.

Reply:
See above comment.

Reply:
OK, I think the base class is basically not needed. The `IAuthorizer` interface should implement the "bind" pattern (like `lambda.Code`) and allow derived classes to access the `RestApi` object. There is no need to store it at the base class.

This is also inline with my comment below about authorizers reporting their own type, which will make the API much cleaner.

So instead of:

```ts
authorizationType: apigateway.AuthorizationType.CUSTOM,
authorizer: auth
```

The eventual API I would envision is:

```ts
authorizer: apigw.Authorizer.custom(lambdaHandler, { /* ... */ });
```

Or for IAM:

```ts
authorizer: apigw.Authorizer.iam()
```

Reply:
> OK, I think the base class is basically not needed. The IAuthorizer interface should implement the "bind" pattern (like lambda.Code) and allow derived classes to access the RestApi object. There is no need to store it at the base class.

🧐 interfaces in ts cannot hold an implementation, can it? `lambda.Code` is an abstract class. 
Also, any change to the interface would be a breaking change to existing customers who might be using `CfnAuthorizer` and plugging it into their REST API Methods.

> my comment below about authorizers reporting their own type

Agree that it would be better for authorizers to report their own type.

> `authorizer: apigw.Authorizer.custom(lambdaHandler, { /* ... */ });`
> `authorizer: apigw.Authorizer.iam()`

Questions to clarify:
1.  The interface we have now is `IAuthorizer`. Are you suggesting we create another class called `Authorizer`?
2. The methods would actually be `apigw.Authorizer.token({ ... })` and `apigw.Authorizer.iam({ ... })` where the parameter to these methods would be `TokenAuthorizerProps` and (potentially) `IAMAuthorizerProps`. I see value in doing this where `apigw.Authorizer.` would generate the list of authorizers we support on the IDE, thus making it a good discoverability experience.